### PR TITLE
[FEAT] Factor cost matrix in dense case

### LINKED ISSUE #49: [ENH] Dense mappings should factor cost matrix by default
As discussed with @bthirion, the cost matrix should be factored via Nystroem approximation for dense mappings as it is the case with the sparse solvers. This should greatly improve time/memory performance even for lower resolution meshes.

### PR DESCRIPTION
**Pull request by pbarbarant**
Solves #49 . Dense mappings now use Nystroem approximation by default for the cost matrix. Closes #49 

- [x] Fixed dense solver/mapping
- [x] Adapted coarse-to-fine procedure
- [ ] Adapt examples in the doc

### CODE REVIEW

File: src/fugw/scripts/coarse_to_fine.py [hunk 1]
```diff
@@ -373,19 +373,17 @@ def fit(
...
-        source_geometry_embeddings[source_sample],
-        source_geometry_embeddings[source_sample],
-        p=2,
+    # Compute coarse geometry embeddings
+    source_coarse_embedding = source_geometry_embeddings[source_sample]
+    target_coarse_embedding = target_geometry_embeddings[target_sample]
+
+    # Normalize embeddings
+    source_coarse_embedding = (
+        source_coarse_embedding / source_coarse_embedding.max()
```
Comment (bthirion) [inline src/fugw/scripts/coarse_to_fine.py:382]: This is a normalization  I guess ? You don't take the absolute value ?

File: src/fugw/solvers/dense.py [hunk 2]
```diff
@@ -128,14 +144,14 @@ def fugw_loss(
         loss_regularization = torch.zeros(1)
         loss = 0
 
-        if alpha != 1 and D is not None:
-            loss_wasserstein = ((D * pi).sum() + (D * gamma).sum()) / 2
+        if alpha != 1 and K1 is not None and K2 is not None:
+            loss_wasserstein = ((K1 @ K2.T) * (pi + gamma)).sum() / 2
```
Comment (bthirion) [inline src/fugw/solvers/dense.py:148]: This is a different formula, right ? Any explanation on the change ?

**Review by bthirion — COMMENTED** [review]
I looks like it broke docs generation

File: src/fugw/scripts/coarse_to_fine.py (hunk 1 above)
Comment (pbarbarant) [reply src/fugw/scripts/coarse_to_fine.py]: Yes my mistake the normalization should have been done with respect to the max norm of the geometry, not the embedding directly. This has been fixed.

File: src/fugw/solvers/dense.py (hunk 2 above)
Comment (pbarbarant) [reply src/fugw/solvers/dense.py:148]: $D \simeq K_1 K_2 ^T$ so this is just a factored expression to be consistent with
```
loss_wasserstein = csr_sum(elementwise_prod_fact_sparse(K1, K2, pi + gamma)) / 2
```
which appears in the sparse solver.

**Review by bthirion — COMMENTED** [review]
OK. There is still a doc building failure...

**Comment by pbarbarant** [comment]
> OK. There is still a doc building failure...

That was expected as the examples using the dense solvers need to receive the embeddings instead of the cost matrix. I'll rewrite them so that it is clear for the user.

**Comment by bthirion** [comment]
I suggest to open an issue if you don't address it in this PR.
The PR LGTM otherwise.
